File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n = len(cost)
        
        def cal_min(idx):
            if idx <=1:
                return 0
            
            if idx in memo:
                return memo[idx]
            
            memo[idx] =  min(cal_min(idx-2)+cost[idx-2], cal_min(idx-1) +cost[idx-1])
            return memo[idx]
        
        memo = {}
        return cal_min(n) # on top        

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int dp(int idx, vector<int>& cost, unordered_map<int, int>& mp){
        // unordered_map<int, int> mp;
        // mp.reserve(cost.size());
        
        if (idx == 0 || idx==1) return 0;
        if (mp.find(idx) != mp.end()) return mp[idx];
        
        mp[idx] = min(dp(idx-2, cost, mp)+cost[idx-2], dp(idx-1, cost, mp)+cost[idx-1]);
        return mp[idx];
    }
    
    int minCostClimbingStairs(vector<int>& cost) {
        unordered_map<int, int> mp;
        mp.reserve(cost.size());
        int ans = dp(cost.size(), cost, mp);
        return ans;
    }
};